### DISCUSSION TITLE
Very large binary size compared to AVR

### DISCUSSION THREAD

**Discussion by Hyratel** [stm32duino/Arduino_Core_STM32#1723] [Q&A] [answered]
been tinkering with some F030F4P6 boards because they were cheap and I was curious and... 16k of flash - I rarely use up half that in my Sketches on AVRs, but I'm routinely running up against the size limit just by the inclusion of Hardware Serial or hardware I2C Wire. it is impossible to use I2C, Serial AND an analogRead() - a single analogRead alone adds 3k to the compiled size, and HWSerial adds 4k, while the presence of any user code puts it at nearly 4k, and doing anything worth doing bumps up to around 8k without any serial activity. this really strikes me as severely off 

I just want to read 10 buttons and an ADC input and send them over Wire

Targetting a Pro Mini compiles to just over 2k
Targetting the F030 is 14.5k

```arduino

#include <Wire.h>
uint8_t buffered[3]= {0,0,0};

void setup()
{
  Wire.begin(9);              
  Wire.onRequest(requestEvent);
}

void loop()
{
  buffered[0] =0;
  for(char i=0; i<8; i++)
  {
    //buttons[0] |= digitalReadFast((PinName)(i+1)) << i;
    buffered[0] |= digitalRead(i+1) << i;
  }
  uint16_t analog0temp = analogRead(A0);
  buffered[1] = lowByte(analog0temp);
  buffered[2] = highByte(analog0temp);
  delay(1);
  
}


void requestEvent()
{
  Wire.write(buffered,3 ); 
}
```

**Comment by fpistm** [comment]
Hi,
Right as the core is generic for all STM32 series and it uses the ST HAL it can consume lot of flash. Moreover STM32 is 32 bit while avr is 8 bit.
You can customize some options to decrease the size:
* By disabling unused HAL module:
https://github.com/stm32duino/wiki/wiki/HAL-configuration#example-of-blinkino-for-nucleo-l031k6-with-32768-bytes-of-flash-and-8192-bytes-of-ram
* Customize pinmap array:
https://github.com/stm32duino/wiki/wiki/Custom-definitions#custom-pinmap-array

**Comment by fpistm** [comment, marked as answer]
As example for your sketch  and generic F030F4Px:

> 
> Sketch uses 14036 bytes (85%) of program storage space. Maximum is 16384 bytes.
> Global variables use 452 bytes (11%) of dynamic memory, leaving 3644 bytes for local variables. Maximum is 4096 bytes.

With:
1. Serial disabled
2. Added file `build_opt.h` with this inside:
```
-DHAL_RTC_MODULE_DISABLED
-DHAL_SPI_MODULE_DISABLED
-DHAL_TIM_MODULE_DISABLED
-DHAL_EXTI_MODULE_DISABLED
```
3. Sketch with PinMap_ADC redefined: 
```C++
#include <Wire.h>
uint8_t buffered[3] = {0, 0, 0};

const PinMap PinMap_ADC[] = {
  {PA_0, ADC1, STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 0, 0)}, // ADC_IN0
  {NC,   NP,   0}
};


void setup()
{
  Wire.begin(9);
  Wire.onRequest(requestEvent);
}

void loop()
{
  buffered[0] = 0;
  for (char i = 0; i < 8; i++)
  {
    buffered[0] |= digitalRead(i + 1) << i;
  }
  uint16_t analog0temp = analogRead(A0);
  buffered[1] = lowByte(analog0temp);
  buffered[2] = highByte(analog0temp);
  delay(1);
}

void requestEvent()
{
  Wire.write(buffered, 3 );
}
```

**Reply by Hyratel** [reply]
it's some savings (I was seeing around 80% flash fill) with the HAL modules disabled and ADC Pinremapped per your `code` , but somehow I got it down to 

> Sketch uses 9592 bytes (58%) of program storage space. Maximum is 16384 bytes.
> Global variables use 384 bytes (9%) of dynamic memory, leaving 3712 bytes for local variables. Maximum is 4096 bytes.

by changing 

> Optimize: Smallest default

to

> Optimize: Smallest with LTO

it seems build options and compiled binary size is not entirely deterministic because I also had some reductions in size by changing some options back and forth between two settings repeatedly

**Reply by fpistm** [reply]
> it seems build options and compiled binary size is not entirely deterministic because I also had some reductions in size by changing some options back and forth between two settings repeatedly

This is due to core.a cache system of Arduino. Best way is to restart IDE to avoid such issue.

**Reply by Hyratel** [reply]
IDE really needs a `Compile Clean` button...
